biz.aQute.repository/src/aQute/bnd/repository/p2/provider/P2Indexer.java

```diff
@@ -130,12 +132,29 @@ private Repository readRepository() throws Exception {
 		P2Impl p2 = new P2Impl(client, this.url, Processor.getExecutor());
 		List<Artifact> artifacts = p2.getArtifacts();
 		List<Promise<Resource>> fetched = new ArrayList<>(artifacts.size());
-		Set<URI> visited = new HashSet<>(artifacts.size());
+		Set<BundleId> visited = new HashSet<>(artifacts.size());
+
+		final Map<BundleId,Resource> existing = new HashMap<>();
+		if (getBridge() != null) {
```

Comment:
You have a circularity problem here. This method is called to create the Repository which the BridgeRepository wraps. So you cannot use the wrapper to create the thing it wraps. So when called from the constructor, there is no BridgeRepository yet (obviously). But when called from `refresh` you are using the outgoing BridgeRepository's to create the data for the incoming BridgeRepository. That hardly seems like a refresh operation. 

So this whole thing seems horribly broken. It is a no-op (`getBridge()` returns `null`) when constructing a P2Indexer object. And when refreshing the object, it uses stale date from the outgoing BridgeRepository. So, yes, the refresh is a whole lot faster at the expense of being potentially wrong since you don't actually refresh. It is not sufficiently safe to assume that the artifact is unchanged even though it has the same bsn+version.